### PR TITLE
Update Foreign Function & Memory API to work with Windows AArch64 ABI

### PR DESCRIPTION
**Pull request by swesonga**
See JEP 424: Foreign Function & Memory API (Preview) - https://openjdk.org/jeps/424

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java [hunk 1]
```diff
@@ -309,8 +332,13 @@ protected void spillStructBox(Binding.Builder bindings, MemoryLayout layout) {
     }
 
     class UnboxBindingCalculator extends BindingCalculator {
-        UnboxBindingCalculator(boolean forArguments) {
+        protected final boolean forArguments;
+        protected final boolean forVariadicFunction;
```
Comment (swesonga) [inline src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java:336]: These should be private

File: src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java [hunk 2]
```diff
@@ -325,6 +353,16 @@ List<Binding> getIndirectBindings() {
         List<Binding> getBindings(Class<?> carrier, MemoryLayout layout) {
             TypeClass argumentClass = TypeClass.classifyLayout(layout);
             Binding.Builder bindings = Binding.builder();
+
+            boolean useIntRegsForFloatingPointArgs = useIntRegsForVariadicFloatingPointArgs() && forVariadicFunction && forArguments;
+
+            // Pass HFA arguments using STRUCT_REGISTER when general purpose registers are being
+            // used to pass floating point arguments. If the HFA is too big to pass entirely in
+            // general purpose registers, then pass it as a struct (i.e. as a STRUCT_REFERENCE).
+            if (useIntRegsForFloatingPointArgs && argumentClass == TypeClass.STRUCT_HFA) {
```
Comment (ahmedmuhsin) [inline src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java:362]: A way to potentially improve the readability of this section would be to extract the parts that you need for this special `STRUCT_HFA` scenario from the bodies of the case statements for `STRUCT_REGISTER` and `STRUCT_REFERENCE` into separate methods. Then you can move the if statement that is defined here to the body of the `STRUCT_HFA` statement and call those extracted methods as necessary. For example:

```
case STRUCT_HFA: {
    ...
    if (useIntRegsForFloatingPointArgs) {
        if (layout.byteSize() <= maxLayoutSizeOfHFAInGeneralPurposeRegs()) {
            methodExtractedFromStructRegister(); 
       }
       else {
            methodExtractedFromStructReference();
       }
    }
    ...
    break;
}
```

Comment (swesonga) [reply src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java:362]: I ended up moving this code during a refactoring so this is no longer applicable. See new PR.

**Comment by swesonga** [comment]
Closing now that https://github.com/openjdk/panama-foreign/pull/754 is open

**Pull request closed without merging**